File: client/modules/extras.ts
import { fatal, trace } from '@trippler/tr_lib/shared'
import { triggerNuiCallback } from '@trippler/tr_lib/client'
import { config } from '../../shared/constants'
import { getState, setState } from '../states'

const UnAvailableExtras: boolean[] = []
const ActiveExtras: boolean[] = []
const InActiveExtras: boolean[] = []

const openDamageCheck = () => {
  if (config.ignoreVehicleState) return false
  if (IsVehicleDamaged(getState.vehicleHandle)) {
    if (getState.remoteBusy) {
      triggerNuiCallback('close', false)
      SetNuiFocus(false, false)
      setState.remoteBusy(false)
      return 1
    } else return 2
  } else return false
}

export const notify = (text: string, type?: string) => {
  SetNotificationTextEntry("STRING")
  AddTextComponentString(text)
  DrawNotification(false, false)
}

export const ToggleDamageCheck = (isExtraOn: boolean, extraNum: number, cb: Function) => {
  if (!config.ignoreVehicleState && IsVehicleDamaged(getState.vehicleHandle)) {
    if (!isExtraOn) {
      notify(config.notify.error.damaged, 'error')
      cb({
        success: false,
        error: "Veh damaged"
      })
    } else {
      SetVehicleExtra(getState.vehicleHandle, extraNum, isExtraOn)
      const newState = !isExtraOn
      trace(`Extra ${extraNum} ${isExtraOn ? "disabled" : "enabled"}`)
      cb({
        success: true,
        extraNum: extraNum,
        isActive: newState
      })
    }
  } else {
    SetVehicleExtra(getState.vehicleHandle, extraNum, isExtraOn)
    const newState = !isExtraOn
    trace(`Extra ${extraNum} ${isExtraOn ? "disabled" : "enabled"}`)
    cb({
      success: true,
      extraNum: extraNum,
      isActive: newState
    })
  }
}

export const checkVehicleExtras = () => {
  for (let i = 1; i <= 12; i++) {
    ActiveExtras[i] = IsVehicleExtraTurnedOn(getState.vehicleHandle, i) ? true : false
    InActiveExtras[i] = !IsVehicleExtraTurnedOn(getState.vehicleHandle, i)
    UnAvailableExtras[i] = !DoesExtraExist(getState.vehicleHandle, i)
  }
}

export const toggleRemote = (): void => {
  if (!getState.vehicleBusy) return
  if (openDamageCheck() == 1) return notify(config.notify.success.closed, 'success')
  if (openDamageCheck() == 2) return notify(config.notify.error.damaged, 'error')
  if (openDamageCheck() == false) {
    setState.remoteBusy(!getState.remoteBusy)
    if (getState.remoteBusy) {
      checkVehicleExtras()
      triggerNuiCallback('open', {
        unAvailableExtras: UnAvailableExtras,
        activeExtras: ActiveExtras,
        inActiveExtras: InActiveExtras,
        sirenOn: true
      })
      setState.remoteBusy(true)
    } else {
      SetNuiFocus(false, false)
      triggerNuiCallback('close', false)
      setState.remoteBusy(false)
    }
  } else fatal('Something unexpected happend')
}

export const toggleCursor = () => {
  setTimeout(() => {
    if (getState.remoteBusy) SetNuiFocus(true, true)
    else SetNuiFocus(false, false)
  }, 200)
}

File: lenixfx/lenix_patrol/client/modules/extras.lua
local shared<const> = require '@trippler/tr_lib/shared'
local lib<const> = require '@trippler/tr_lib/client'
local config<const> = require 'shared/constants/index'
local state<const> = require 'client/states'

local UnAvailableExtras = {}
local ActiveExtras = {}
local InActiveExtras = {}

local function openDamageCheck()
  if config.ignoreVehicleState then return false end
  if IsVehicleDamaged(state.getState.vehicleHandle) then
    if state.getState.remoteBusy then
      lib.triggerNuiCallback('close', false)
      SetNuiFocus(false, false)
      state.setState.remoteBusy(false)
      return 1
    else 
      return 2 
    end
  else 
    return false 
  end
end

local function notify(text, type)
  SetNotificationTextEntry("STRING")
  AddTextComponentString(text)
  DrawNotification(false, false)
end

local function ToggleDamageCheck(isExtraOn, extraNum, cb)
  if not config.ignoreVehicleState and IsVehicleDamaged(state.getState.vehicleHandle) then
    if not isExtraOn then
      notify(config.notify.error.damaged, 'error')
      cb({ success = false, error = "Veh damaged" })
    else
      SetVehicleExtra(state.getState.vehicleHandle, extraNum, isExtraOn and 1 or 0)
      local newState<const> = not isExtraOn
      shared.trace(('Extra %d %s'):format(extraNum, isExtraOn and "disabled" or "enabled"))
      cb({ success = true, extraNum = extraNum, isActive = newState })
    end
  else
    SetVehicleExtra(state.getState.vehicleHandle, extraNum, isExtraOn and 1 or 0)
    local newState<const> = not isExtraOn
    shared.trace(('Extra %d %s'):format(extraNum, isExtraOn and "disabled" or "enabled"))
    cb({ success = true, extraNum = extraNum, isActive = newState })
  end
end

local function checkVehicleExtras()
  local veh<const> = state.getState.vehicleHandle
  for i = 1, 12 do
    ActiveExtras[i] = IsVehicleExtraTurnedOn(veh, i) == 1
    InActiveExtras[i] = not ActiveExtras[i]
    UnAvailableExtras[i] = not DoesExtraExist(veh, i)
  end
end

local function toggleRemote()
  if not state.getState.vehicleBusy then return end
  local check<const> = openDamageCheck()
  
  if check == 1 then return notify(config.notify.success.closed, 'success') end
  if check == 2 then return notify(config.notify.error.damaged, 'error') end
  
  if check == false then
    state.setState.remoteBusy(not state.getState.remoteBusy)
    if state.getState.remoteBusy then
      checkVehicleExtras()
      lib.triggerNuiCallback('open', {
        unAvailableExtras = UnAvailableExtras,
        activeExtras = ActiveExtras,
        inActiveExtras = InActiveExtras,
        sirenOn = true
      })
      state.setState.remoteBusy(true)
    else
      SetNuiFocus(false, false)
      lib.triggerNuiCallback('close', false)
      state.setState.remoteBusy(false)
    end
  else 
    shared.fatal('Something unexpected happened') 
  end
end

local function toggleCursor()
  SetTimeout(200, function()
    if state.getState.remoteBusy then SetNuiFocus(true, true)
    else SetNuiFocus(false, false) end
  end)
end

return {
  ToggleDamageCheck = ToggleDamageCheck,
  checkVehicleExtras = checkVehicleExtras,
  toggleRemote = toggleRemote,
  toggleCursor = toggleCursor,
  notify = notify
}